Framework/Source/Graphics/Program/ProgramVars.cpp

```diff
@@ -0,0 +1,222 @@
+    {
+        return mpGlobalBlock->getRawBuffer(name);
+    }
+
+    TypedBufferBase::SharedPtr ProgramVars::getTypedBuffer(const std::string& name) const
+    {
+        return mpGlobalBlock->getTypedBuffer(name);
+   }
+
+    StructuredBuffer::SharedPtr ProgramVars::getStructuredBuffer(const std::string& name) const
+    {
+        return mpGlobalBlock->getStructuredBuffer(name);
+    }
+
+    bool ProgramVars::setSampler(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex, const Sampler::SharedPtr& pSampler)
+    {
+        return mpGlobalBlock->setSampler(regSpace, baseRegIndex, arrayIndex, pSampler);
+    }
+
+    bool ProgramVars::setSampler(const std::string& name, const Sampler::SharedPtr& pSampler)
+    {
+        return mpGlobalBlock->setSampler(name, pSampler);
+    }
+
+    Sampler::SharedPtr ProgramVars::getSampler(const std::string& name) const
+    {
+        return mpGlobalBlock->getSampler(name);
+    }
+
+    Sampler::SharedPtr ProgramVars::getSampler(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex) const
+    {
+        return mpGlobalBlock->getSampler(regSpace, baseRegIndex, arrayIndex);
+    }
+
+    ShaderResourceView::SharedPtr ProgramVars::getSrv(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex) const
+    {
+        return mpGlobalBlock->getSrv(regSpace, baseRegIndex, arrayIndex);
+    }
+
+    UnorderedAccessView::SharedPtr ProgramVars::getUav(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex) const
+    {
+        return mpGlobalBlock->getUav(regSpace, baseRegIndex, arrayIndex);
+    }
+
+    bool ProgramVars::setTexture(const std::string& name, const Texture::SharedPtr& pTexture)
+    {
+        return mpGlobalBlock->setTexture(name, pTexture);
+    }
+
+    Texture::SharedPtr ProgramVars::getTexture(const std::string& name) const
+    {
+        return mpGlobalBlock->getTexture(name);
+    }
+
+    bool ProgramVars::setSrv(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex, const ShaderResourceView::SharedPtr& pSrv)
+    {
+        return mpGlobalBlock->setSrv(regSpace, baseRegIndex, arrayIndex, pSrv);
+    }
+
+    bool ProgramVars::setUav(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex, const UnorderedAccessView::SharedPtr& pUav)
+    {
+        return mpGlobalBlock->setUav(regSpace, baseRegIndex, arrayIndex, pUav);
+    }
+
+    template<bool forGraphics>
+    bool applyProgramVarsCommon(const ProgramVars* pVars, ParameterBlock* pGlobalBlock, CopyContext* pContext, bool bindRootSig)
+    {
+        if (bindRootSig)
+        {
+            if (forGraphics)
+            {
+                pVars->getRootSignature()->bindForGraphics(pContext);
+            }
+            else
+            {
+                pVars->getRootSignature()->bindForCompute(pContext);
+            }
+        }
+
+        if (pGlobalBlock->prepareForDraw(pContext, pVars->getRootSignature().get()) == false) return false;
+
```

Comment:
Hmm... I was almost expecting that this logic would involve calling `prepareForDraw` on the main/global block (which you do here) which would then go on to recursively call `prepareForDraw` on any blocks that got bound to global-scope parameter-block parameters.

Maybe you'll be able to walk me through the sequence of how things get bound here. Ideally I'd expect the main parameter block to be handled similarly to all the others... right?

Reply:
The global block doesn't store the rest of the blocks.
`ProgramVars` stores an array of blocks (not implemented yet). We need to call `prepareForDraw` on each one of them, and they will recursively call it on internal blocks.
Actually, `prepareForDraw` is not required. I should store a dirty flag inside the block and call `prepareForDraw` internally when the user calls `getRootSets()`